sw/device/lib/base/hardened_status.h

```diff
@@ -0,0 +1,60 @@
+// Copyright lowRISC contributors.
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+
+#ifndef OPENTITAN_SW_DEVICE_LIB_BASE_HARDENED_STATUS_H_
+#define OPENTITAN_SW_DEVICE_LIB_BASE_HARDENED_STATUS_H_
+
+/**
+ * @file
+ * @brief Hardened handling of status codes.
+ */
+
+#include "sw/device/lib/base/hardened.h"
+#include "sw/device/lib/base/status.h"
+
+#ifdef __cplusplus
+extern "C" {
+#endif  // __cplusplus
+
+/**
+ * Hardened variant of `OK_STATUS` .
+ *
+ * This passes `kHardenedBoolTrue` as the status code argument, for extra bits
+ * of redundancy in `HARDENED_TRY` and others.
+ */
+#define HARDENED_OK_STATUS OK_STATUS(kHardenedBoolTrue)
```

Comment:
You might want to have more strict status code creation requirements for hardened status codes.

In status.h, all of the status creating macros (`OK_STATUS`, `INVALID_ARGUMENT`, etc) go through the `status_create` function in status.c to glue all of the components of a status together (error/ok bit, module_id, code, argument).  In the event that the particular translation unit (e.g C file) invoking the macro does not define a local `kModuleId`, the `status_create` function reads `__FILE__` and makes up a module ID from the filename.

This may be of concern for two reasons:
1. It involves a loop doing conventional C-style string processing (finding the terminator byte, backing up to the last path component, etc).
2. In the use cases for hardened codes, you may not want a function call to have to assemble the status code together.

_If_ you were to require that so-called hardened status code were always created in modules that had a module-id (created with the `DECLARE_MODULE_ID` macro from internal/status.h), you could make status code creation purely out of constants rather than delegating to a function call:


If I'm remembering my bitfield ordering correctly:
```
#define HARDENED_OK_STATUS ( (status_t){.value = kHardenedBoolTrue})
#define HARDENED_INVALID_ARGUMENT ((status_t){.value = 
    0x80000000 | kModuleId | ((__LINE__ & 0x7ff) << 5) | kInvalidArgument })
(etc)
```

It would be ok to do this additional work in a follow-up PR.

Reply:
Thanks for pointing this out! For cryptolib, I need to eventually map these errors onto a much smaller set of errors from the cryptolib API. So I was planning to introduce another cryptolib-specific layer of definitions:
```c
#define OTCRYPTO_OK HARDENED_OK_STATUS
#define OTCRYPTO_RECOV_ERR STATUS_CREATE(kAborted, __LINE__)
#define OTCRYPTO_FATAL_ERR STATUS_CREATE(kFailedPrecondition, __LINE__)
#define OTCRYPTO_BAD_ARGS STATUS_CREATE(kInvalidArgument, __LINE__)
#define OTCRYPTO_ASYNC_INCOMPLETE STATUS_CREATE(kUnavailable, __LINE__)
```

In addition, I'd have a `status_interpret()` that would convert a `status_t` into a `crypto_status_t`. The invariant here is that e.g. `OTCRYPTO_FATAL_ERR` produces a type of `status_t` that will definitely be interpreted as `kCryptoStatusFatalError`. Only a hardened `OK` gets interpreted as an OK cryptolib status. In the cryptolib, I'd only use the crypto-specific error types (although non-cryptolib-specific errors would still typecheck).

So, with that context: would it make sense to put the hardcoded `kModuleId` in the cryptolib-specific layer instead? e.g.
```c
#define OTCRYPTO_RECOV_ERR ((status_t){.value = 
    0x80000000 | kModuleId | ((__LINE__ & 0x7ff) << 5) | kAborted });
```

Reply:
Will implement this in one layer or another in the next PR :slightly_smiling_face: